File: apps/app2/src/traffic_generator.py
#!/usr/bin/env python3

import json
import os
import random
import requests
import time
import threading
from datetime import datetime
from typing import List, Dict

class TrafficGenerator:
    def __init__(self):
        self.target_url = os.getenv("TARGET_URL", "http://app2-service:8000")
        self.request_interval = int(os.getenv("REQUEST_INTERVAL", "4"))
        self.error_rate = float(os.getenv("ERROR_RATE", "0.12"))
        
        self.endpoints = [
            {"path": "/health", "weight": 0.3, "method": "GET"},
            {"path": "/api/data", "weight": 0.4, "method": "GET"},
            {"path": "/api/compute", "weight": 0.2, "method": "GET"},
            {"path": "/api/database", "weight": 0.1, "method": "GET"}
        ]
        
        self.session = requests.Session()
        self.session.timeout = 10
    
    def log_message(self, level: str, message: str, **kwargs):
        log_entry = {
            "timestamp": datetime.utcnow().isoformat(),
            "level": level,
            "service": "app2-traffic-generator",
            "message": message,
            **kwargs
        }
        print(json.dumps(log_entry))
    
    def select_endpoint(self) -> Dict:
        """Selecciona un endpoint basado en los pesos configurados"""
        r = random.random()
        cumulative_weight = 0
        
        for endpoint in self.endpoints:
            cumulative_weight += endpoint["weight"]
            if r <= cumulative_weight:
                return endpoint
        
        return self.endpoints[0]  # fallback
    
    def make_request(self, endpoint: Dict):
        url = f"{self.target_url}{endpoint['path']}"
        
        try:
            start_time = time.time()
            response = self.session.request(endpoint["method"], url)
            duration = time.time() - start_time
            
            status = "success" if response.status_code < 400 else "error"
            
            self.log_message(
                "info",
                f"Request completed",
                endpoint=endpoint["path"],
                method=endpoint["method"],
                status_code=response.status_code,
                duration_ms=round(duration * 1000, 2),
                status=status
            )
            
        except requests.exceptions.RequestException as e:
            self.log_message(
                "error",
                f"Request failed: {str(e)}",
                endpoint=endpoint["path"],
                method=endpoint["method"],
                error_type=type(e).__name__
            )
    
    def generate_burst_traffic(self):
        """Genera ráfagas de tráfico para simular picos de carga"""
        while True:
            # Esperar entre 30-120 segundos para la próxima ráfaga
            wait_time = random.randint(30, 120)
            time.sleep(wait_time)
            
            # Generar ráfaga de 5-15 requests
            burst_size = random.randint(5, 15)
            
            self.log_message(
                "info",
                f"Generating traffic burst",
                burst_size=burst_size
            )
            
            threads = []
            for _ in range(burst_size):
                endpoint = self.select_endpoint()
                thread = threading.Thread(target=self.make_request, args=(endpoint,))
                threads.append(thread)
                thread.start()
                
                # Pequeña pausa entre requests de la ráfaga
                time.sleep(random.uniform(0.1, 0.5))
            
            # Esperar que terminen todos los threads
            for thread in threads:
                thread.join()
    
    def generate_regular_traffic(self):
        """Genera tráfico regular y constante"""
        self.log_message(
            "info",
            "Traffic generator started",
            target_url=self.target_url,
            request_interval=self.request_interval
        )
        
        while True:
            try:
                # Número de requests simultáneos (1-3)
                concurrent_requests = random.randint(1, 3)
                
                threads = []
                for _ in range(concurrent_requests):
                    endpoint = self.select_endpoint()
                    thread = threading.Thread(target=self.make_request, args=(endpoint,))
                    threads.append(thread)
                    thread.start()
                    
                    # Pausa muy pequeña entre iniciar threads
                    if concurrent_requests > 1:
                        time.sleep(random.uniform(0.1, 0.3))
                
                # Esperar que terminen
                for thread in threads:
                    thread.join()
                
                # Esperar hasta el próximo ciclo
                base_interval = self.request_interval
                jitter = random.uniform(-1, 1)  # ±1 segundo de jitter
                sleep_time = max(1, base_interval + jitter)
                time.sleep(sleep_time)
                
            except KeyboardInterrupt:
                self.log_message("info", "Traffic generator stopping")
                break
            except Exception as e:
                self.log_message(
                    "error",
                    f"Unexpected error in traffic generator: {str(e)}",
                    error_type=type(e).__name__
                )
                time.sleep(5)  # Breve pausa antes de continuar
    
    def run(self):
        """Ejecuta ambos generadores de tráfico en paralelo"""
        # Iniciar el generador de ráfagas en un thread separado
        burst_thread = threading.Thread(target=self.generate_burst_traffic, daemon=True)
        burst_thread.start()
        
        # Ejecutar el tráfico regular en el thread principal
        self.generate_regular_traffic()

if __name__ == "__main__":
    generator = TrafficGenerator()
    generator.run()

File: apps/app1/cmd/traffic-generator/traffic-generator.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"os"
	"time"
)

type TrafficConfig struct {
	TargetURL       string `json:"target_url"`
	RequestInterval int    `json:"request_interval_seconds"`
	ErrorRate       float32 `json:"error_rate"`
}

func loadConfig() TrafficConfig {
	config := TrafficConfig{
		TargetURL:       "http://app1-service:8080",
		RequestInterval: 5,
		ErrorRate:       0.1,
	}
	
	if url := os.Getenv("TARGET_URL"); url != "" {
		config.TargetURL = url
	}
	
	return config
}

func makeRequest(url string, endpoint string) {
	client := &http.Client{
		Timeout: 10 * time.Second,
	}
	
	resp, err := client.Get(url + endpoint)
	if err != nil {
		log.Printf("Error making request to %s%s: %v", url, endpoint, err)
		return
	}
	defer resp.Body.Close()
	
	status := "success"
	if resp.StatusCode >= 400 {
		status = "error"
	}
	
	logEntry := map[string]interface{}{
		"timestamp": time.Now().Format(time.RFC3339),
		"level":     "info",
		"service":   "app1-traffic-generator",
		"message":   fmt.Sprintf("Request to %s - Status: %d", endpoint, resp.StatusCode),
		"endpoint":  endpoint,
		"status":    status,
	}
	
	logJSON, _ := json.Marshal(logEntry)
	fmt.Println(string(logJSON))
}

func generateTraffic() {
	config := loadConfig()
	
	endpoints := []string{"/health", "/data", "/slow"}
	weights := []float32{0.5, 0.4, 0.1} // Probabilidades relativas
	
	logEntry := map[string]interface{}{
		"timestamp":  time.Now().Format(time.RFC3339),
		"level":      "info",
		"service":    "app1-traffic-generator",
		"message":    "Traffic generator started",
		"target_url": config.TargetURL,
	}
	
	logJSON, _ := json.Marshal(logEntry)
	fmt.Println(string(logJSON))
	
	ticker := time.NewTicker(time.Duration(config.RequestInterval) * time.Second)
	defer ticker.Stop()
	
	for {
		select {
		case <-ticker.C:
			// Seleccionar endpoint basado en pesos
			r := rand.Float32()
			var endpoint string
			
			if r < weights[0] {
				endpoint = endpoints[0]
			} else if r < weights[0]+weights[1] {
				endpoint = endpoints[1]
			} else {
				endpoint = endpoints[2]
			}
			
			// Generar múltiples requests para simular carga
			numRequests := 1 + rand.Intn(3) // 1-3 requests
			
			for i := 0; i < numRequests; i++ {
				go makeRequest(config.TargetURL, endpoint)
				
				// Pequeña pausa entre requests
				if i < numRequests-1 {
					time.Sleep(time.Duration(100+rand.Intn(400)) * time.Millisecond)
				}
			}
		}
	}
}

func main() {
	// Seed para randomización
	rand.Seed(time.Now().UnixNano())
	
	generateTraffic()
}